### DISCUSSION TITLE
Allow properties to be nullable

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#13816] [Features and ideas]
Currently when using numeric or decimal properties ModelsBuilder generate these as `int` og `decimal`, but there doesn't seem to be a way to defined it as nullable, although the property in backoffice is allowed to be empty.

Maybe the properties should have a setting "Allow empty" or similar, which will generate to properties as `int?` and `decimal?` instead and it would be easier to handle a fallback value yourself as `0` sometimes is a valid value, but can be confusing if some properties are empty while others have the value `0`.

It can of course use textstring instead `string`, but then one will not get the benefits of the numeric input. An alternative is to use textstring + regex on the property in backoffice to ensure a numeric value, but the generates type is still `string`.

It seems the be the same for textstring which always is generates as `string` and not `string?`.

Something like this would be great :)

![image](https://user-images.githubusercontent.com/2919859/217807435-f2c2cd73-9556-4a88-bfbb-da509d015151.png)

**Comment by bergmania** [comment]
One could argue if a field is not mandatory, it should be nullable?

Thereby we do not need a new setting.

Overall I like the idea about making ModelsBuilder Nullable aware.

**Reply by bjarnef** [reply]
Possible, but not sure if there are use-cases where a property is not mandatory, but one want it not to nullable in code... but then the business logic in code can fallback to a default value.

I guess using the exiting mandatory setting would be a breaking change, but a new setting may not require a breaking change?

Sometimes we also change the mandatory setting later, but if the nullable property is based on this I guess it may require some code changes when generating models. Mostly we use `SourceCodeManual` mode, so not sure how it would affect the other ModelsBuilder modes.

**Reply by bergmania** [reply]
Changing doctype configuration will almost always require some code changes?

Code wise, It just seems weird if a property do not allow to be empty, but is not mandatory? 
For reference-types is default value would then be `null`.

**Reply by bjarnef** [reply]
Yes, I guess it makes sense.. just until now validation hasn't effected models generations.

For now it is also possible a property is not mandatory, but the e.g. a numeric or decimal property has 0 as default value for `int` / `decimal`, which sometimes could be an issue when 0 is a valid value, e.g. it may be hard to tell if a product is actually free or the product just hasn't assigned a price, where one may want to hide the "add to basket" button and quantity field on product page. A solution here is to use textbox instead, a custom property editor or an additional property or rules to decide whether the product is free or just hasn't a price.

**Reply by JasonElkin** [reply]
I'm here for this!

Value types are straightforward, and IMO the default approach, where an empty field can return 0 is misleading - in situations described exactly as above.

I have written a package to try and address this: https://github.com/JasonElkin/Emptiness

I would love it if this was how Umbraco behaved OoTB.

WRT determining if reference types are nullable, I'm not sure a toggle in the back office is the best approach.There are packages, for example Contentment, where the nullability of the property depends on more complex configuration.

**Reply by abjerner** [reply]
I will make the argument that maybe this should be controlled at the data type level instead of at the property level. Eg. a value converter could have a fallback value if the property doesn't have a value.

The benefit of this will be that the developer of the property editor is still in control - eg. via the value converter. Downside will be if the property editor developer doesn't handle this scenario.

Ideally all Umbraco property editors (or at least those that make sense) should have a **Nullable** option on the data type. I know this creates some extra work handling it on the data type level, but like mentioned, I think it gives some more control.

<hr />

For instance, this is how I've handled it for my custom textbox property editor:

![image](https://user-images.githubusercontent.com/3634580/223386049-da5cc1e2-c75f-4ccb-82eb-8b59750533f5.png)

If I configure a fallback value in the data type, the value is never `null`. If I don't configure a fallback and don't enable the **Nullable** option, I get `string.Empty`, where as with the **Nullable** option enabled, I get `null`.

<hr />

Regarding ModelsBuilder, there currently isn't a good way for a value converter to indicate whether a refence type is nullable, as `string?` is essentially the same as `string`, but with an extra flag. `System.Type` doesn't support this flag.

I think the solution could be to create a `NullableType` class that extends `System.Type`, and then has an `IsNullable` property. If this is the case, ModelsBuilder could call the `GetPropertyValueType` method on the value converter, and then check whether the returned type is an instance of `NullableType`. Or perhaps the name of the class should be future proofed for additional settings, so maybe `ModelsBuilderType` or `UmbracoType` instead.

This is something I've discussed with a few members of the community already - eg. for adding this to my custom Models Builder package. But I haven't had the time to investigate this further. And it would probably be best if this type exists in Umbraco.
